controllers/resource_controller.go

```diff
@@ -0,0 +1,525 @@
+	versionCache pkgaddon.VersionCacheClient
+}
+
+func NewResourceController(mgr manager.Manager, stopChan <-chan struct{}, addonversioncache pkgaddon.VersionCacheClient) (controller.Controller, error) {
+	r := &resourceReconcile{
+		client:       mgr.GetClient(),
+		log:          ctrl.Log.WithName(resourceControllerName),
+		versionCache: addonversioncache,
+	}
+
+	c, err := controller.New(resourceControllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.Service{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.ServiceAccount{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.Namespace{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.ConfigMap{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &rbac_v1.ClusterRole{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &rbac_v1.ClusterRoleBinding{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.DaemonSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.ReplicaSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.StatefulSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &batchv1.Job{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &batch_v1beta1.CronJob{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *resourceReconcile) enqueueRequestWithAddonLabel() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		var reqs = make([]reconcile.Request, 0)
+		var labels = a.GetLabels()
+		if name, ok := labels["app.kubernetes.io/name"]; ok && strings.TrimSpace(name) != "" {
+			if ok, v := r.versionCache.HasVersionName(name); ok {
+				reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      v.Name,
+					Namespace: v.Namespace,
+				}})
+			}
+		}
+		return reqs
+	})
+}
+
+// Reconcile addon dependent resources
+func (r *resourceReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.log.Info("reconciling", "request", req)
+
+	namespace := &v1.Namespace{}
+	if err := r.client.Get(ctx, req.NamespacedName, namespace); err == nil {
```

Comment:
The repeated client.Get calls seem very inefficient, it would be better to generate Typed Resource controllers instead e.g., DeploymentResourceController.

Reply:
Why would we call api multiple times when we could use Dynamic client to get unstructured (runtime.Object) then determine GVK and then convert it to type?
```
scheme.Scheme.Convert(obj, x, 0)
```

There may be other ways to create versioned/typed object from unstructured but I think that would be more effiecient than re-calling api.

Reply:
If you insist on typed client get then the only way is to create multiple sub resource reconcilers or you use dynamic client and type conversion.

Reply:
> Why would we call api multiple times when we could use Dynamic client to get unstructured (runtime.Object) then determine GVK and then convert it to type?
> 
> ```
> scheme.Scheme.Convert(obj, x, 0)
> ```
> 
> There may be other ways to create versioned/typed object from unstructured but I think that would be more effiecient than re-calling api.

None of the addon resource objects are unstructured objects (service, deployment,  cronjob, job, statefulset, replicaset, ), they are well-defined typed objects, and are not custom resources which could leverage dynamic client and unstructured objects.

Reply:
> If you insist on typed client get then the only way is to create multiple resource reconcilers or you use dynamic client and type conversion.

Can you clarify what you mean "create multiple resource reconcilers" ?
Or, use dynamic client and type conversion ?

Reply:
I am not sure how dynamic client can make the check more efficient, following is code snippet to use dynamic client which would be similar as typed resources, and the GVR has to switch to all target resources

`gvr := schema.GroupVersionResource{
		Group:    "",
		Version:  "",
		Resource: "Namespaces",
	}
	un, err := r.dynClient.Resource(gvr).Get(ctx, req.NamespacedName.Name, metav1.GetOptions{})

	if err != nil && un != nil{
			gvk := (*un).GroupVersionKind()
			if gvk.Kind == "Namespace" {
				namespace := &v1.Namespace{}
				err = scheme.Scheme.Convert(un, namespace, 0)
				if err == nil {
					r.log.Info(fmt.Sprintf("processing namespace %s", namespace.Name))
					err = r.handleNamespace(ctx, namespace)
					if err != nil {
						return reconcile.Result{}, fmt.Errorf("failed handling namespace %#v", err)
					}
					return reconcile.Result{}, nil
				}
				return reconcile.Result{}, fmt.Errorf("failed converting to namespace %#v", err)				}
			}`

Reply:
In that case then we must have Reconcile per type, it's just how kubebuilder is designed. That can be done by generating the controllers or using cli e.g., `kubebuilder create api --group "apps" --version v1 --kind Deployment --controller --resource=false`

Reply:
I dont see any function/maintenance/performance/scalability using current flow reconcile resources objects.

But I do like to understand your insistent of a different way, how about I create an issue and follow your idea later on, and implement it as an improvement

And we close this PR as current implementation.

Reply:
I don't agree, in the worst case scenario for `Daemonset` you'll make 9 extraneous calls to API before you reconcile that object. We should rethink this approach if we don't want to implement typed  controller reconcilers then another approach should be found this one is not optimal.

Reply:
@kevdowney followin is the data within controller-runtime.Request that passed to the controllerRuntime.Reconcile function

// Request contains the information necessary to reconcile a Kubernetes object.  This includes the
// information to uniquely identify the object - its Name and Namespace.  It does NOT contain information about
// any specific Event or the object contents itself.

Reconcile only has Namespace and Name information - without object contents itself. ATM, I am not aware any magic tools that can skip the Type Check in the loop. Need think about it.

Reply:
Each controller for each type resource can reduce the "reconcile resource type check" call, but, it also introduces 
- controller.manager controller context swtich cost, which significantly impact addon-controller efficiency 
- impacts operation efficiency, 13+ typed controller added in

Above framework management comparing to 13 Client API call within one function  brings more debug/maintainable cost.